components/places/src/storage/bookmarks/public_node.rs

```diff
@@ -610,7 +657,16 @@ mod test {
             );
             std::thread::sleep(std::time::Duration::from_millis(10));
         }
-        let bmks = recent_bookmarks(&conns.read, 3)?;
+
+        append_invalid_bookmark(
+            &conns.write,
+            &BookmarkRootGuid::Unfiled.guid(),
+            "invalid url",
+            "badurl",
+        );
+
+        // The limit applies before we filter the invalid bookmark, so ask for 4.
```

Comment:
This makes sense to me! 👍🏻 If invalid bookmarks are a common occurrence, something is really wrong, so assuming most of them are valid and not pushing complicated validation logic into the SQL is a great approach.

Will pagination do the right thing as well?

Reply:
> Will pagination do the right thing as well?

None of the functions I touched support pagination, so I think we'll be OK. IIRC, that's going to be more of a concern for history.